### PR TITLE
Agent.final?

### PR DESCRIPTION
**Pull request by ASeegull**
1. Agent id is read from flag at the start of app
2. Removed post request to server for authentification.
3. Adds validation of agent with privat key
4. Adds function to send message to the specific agent
5. Checks if previous state of agent exists
6. Writes current state of agent to yaml file on update

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -1,95 +1,86 @@
...
+	path := "/agentState/" + id + ".yaml"
+	if _, err := os.Stat(path); err != nil {
+		return nil, err
 	}
 
-	body, err := ioutil.ReadAll(response.Body)
+	yamlFile, err := ioutil.ReadFile(path)
 	if err != nil {
-		return nil, errors.Wrapf(err, "could not read response.body")
+		return nil, errors.Wrapf(err, "could not read yaml file %", path)
```
Comment (AndriiPryslupskyi) [inline agent/agent.go:65]: Here must be just "%"?

File: agent/agent.go [hunk 2]
```diff
@@ -1,95 +1,86 @@
...
 	}
-	defer response.Body.Close()
 
-	setup := &pb.Setup{}
-	if err = proto.Unmarshal(body, setup); err != nil {
-		return nil, errors.Wrapf(err, "could not unmarshal setup")
+	if err = yaml.Unmarshal(yamlFile, state); err != nil {
+		return nil, errors.Wrapf(err, "could not decode config file %", path)
```
Comment (anastasiia-skliar) [inline agent/agent.go:69]: %s ?

File: server/database/mongo.go [hunk 3]
```diff
@@ -53,32 +58,37 @@ func SaveState(agentInfo *pb.Agentstate) error {
...
 		go func(val *string) {
 			var agent FoodAgent
-			defer wg.Done()
+			defer syn.Done()
 			if err := c.Find(bson.M{"agentid": val}).One(&agent); err != nil {
-				mutex.Lock()
-				notFound = append(notFound, *val)
-				mutex.Unlock()
+				{
+					syn.Lock()
```
Comment (anastasiia-skliar) [inline server/database/mongo.go:77]: syn.Lock() is not preventing from writing to map notFound

File: agent/agent.go [hunk 4]
```diff
@@ -108,55 +99,94 @@ func streamAgentState(ctx context.Context, agent *controls, messages chan []byte
...
-				if types == websocket.TextMessage || err == io.ErrUnexpectedEOF {
-					log.Info(message)
+				if types == websocket.TextMessage {
+					log.Info(string(message))
 					continue
 				}
-				if types == websocket.CloseMessage {
-					return
+
+				if err != nil {
```
Comment (AndriiPryslupskyi) [inline agent/agent.go:128]: Maybe before type checking must be error checking?

File: agent/agent.go [hunk 5]
```diff
@@ -108,55 +99,94 @@ func streamAgentState(ctx context.Context, agent *controls, messages chan []byte
...
+					log.Errorf("failed to read message from server %s", err)
+					break
 				}
-				if err = agent.setup.UnmarshalToStruct(message); err != nil {
-					log.Errorf("failed unmarshal messages: %s", err)
-					return
+
+				setup := &pb.Setup{}
+				if err = setup.UnmarshalToStruct(message); err != nil {
+					break
```
Comment (AndriiPryslupskyi) [inline agent/agent.go:135]: Error logging

File: server/database/mongo.go [hunk 6]
```diff
@@ -53,32 +58,37 @@ func SaveState(agentInfo *pb.Agentstate) error {
...
-			foods = append(foods, FoodInfo{agent.Product, agent.Weight, agent.StateExpires, ""})
-			mutex.Unlock()
+
+			{
+				syn.Lock()
+				foods = append(foods, FoodInfo{agent.Product, agent.Weight, agent.StateExpires, ""})
+				syn.Unlock()
+			}
+
 		}(&containersID[i])
```
Comment (AndriiPryslupskyi) [inline server/database/mongo.go:89]: You can use value from range and skip index.

File: server/database/mongo.go [hunk 7]
```diff
@@ -53,32 +58,37 @@ func SaveState(agentInfo *pb.Agentstate) error {
...
 			var agent FoodAgent
-			defer wg.Done()
+			defer syn.Done()
 			if err := c.Find(bson.M{"agentid": val}).One(&agent); err != nil {
-				mutex.Lock()
-				notFound = append(notFound, *val)
-				mutex.Unlock()
+				{
+					syn.Lock()
+					notFound = append(notFound, *val)
```
Comment (AndriiPryslupskyi) [inline server/database/mongo.go:78]: You should logging error.

File: server/database/postgres.go [hunk 8]
```diff
@@ -92,11 +115,26 @@ func CheckAgent(idUser string, idAgent string) (bool, error) {
 	return agent.ID != "", nil
 }
 
+// GetAgentOwner returns user ID registered for particular agent
+func GetAgentOwner(agentID string) (string, error) {
+	agent := Agent{}
+	err := db.Where("agents.id = ?", agentID).Find(&agent).Error
+	// If agent has no user id, thats ok, user will sign up soon
+	if err == gorm.ErrRecordNotFound {
```
Comment (AndriiPryslupskyi) [inline server/database/postgres.go]: Why do you make additional error checking?

File: server/database/postgres.go [hunk 9]
```diff
@@ -92,11 +115,26 @@ func CheckAgent(idUser string, idAgent string) (bool, error) {
...
+	}
+	return agent.UserID, nil
+}
+
 //RegisterNewAgent adds a new agent to user, returns nil if adding was successful
-func RegisterNewAgent(idUser string, idAgent string) error {
-	agent := Agent{ID: idAgent, UserID: idUser}
+func RegisterNewAgent(idAgent string) error {
+	agent := Agent{ID: idAgent}
 	rows := db.Create(&agent).RowsAffected
```
Comment (AndriiPryslupskyi) [inline server/database/postgres.go:136]: You save agent without user.

File: server/handler.go [hunk 10]
```diff
@@ -41,122 +53,113 @@ func checkSession(h http.HandlerFunc) http.HandlerFunc {
...
 	}
 
-	log.Printf("New websocket connect with %s", r.Host)
+	if agentState.Token == database.PublicToken {
+		if err = database.RegisterNewAgent(agentState.AgentID); err != nil {
+			return "", err
+		}
+	} else {
+
+	}
```
Comment (AndriiPryslupskyi) [inline server/handler.go:104]: Empty else.

File: agent/agent.go [hunk 11]
```diff
@@ -108,55 +99,94 @@ func streamAgentState(ctx context.Context, agent *controls, messages chan []byte
 				return
 			}
 			log.Info("Agent state sent")
+		case <-messages:
+			msg := <-messages
+			ticker.Stop()
+			state := updateAgent(agent.ID, msg)
+			err = saveAgent(state)
+			ticker = time.NewTicker(time.Second * time.Duration(agent.setup.Heartbeat))
 		}
```
Comment (AndriiPryslupskyi) [inline agent/agent.go:108]: error skipping

File: server/database/postgres.go (hunk 9 above)
Comment (ASeegull) [reply server/database/postgres.go:136]: yes, when agent connects, there is no entry for user in database. We should discuss it tomorrow

File: server/handler.go (hunk 10 above)
Comment (ASeegull) [reply server/handler.go:104]: suspicious..... :D

**Pull request closed without merging**